src/main/scala/metadata.scala

```diff
@@ -0,0 +1,83 @@
+package suffuse
+
+/** AKey is an attribute key: a Metadata map key which carries
+ *  both the type of the corresponding value, and a value to
+ *  provide when there is no value in the map. This as opposed
+ *  to wrapping everything in sight in Option.
+ */
+final class AKey[A](description: String, empty: => A) {
+  def emptyValue: A = empty
+  override def toString = description
+}
+
+/** Attr is a dependently typed key/value pair.
+ */
+trait Attr extends Any {
+  type Type
+  def key: AKey[Type]
+  def value: Type
+
+  def pair                               = key -> value
+  def matches[A](implicit that: AKey[A]) = key == that
+  def to_s                               = s"$key: $value"
+}
+
+/** The simple implementation of Attr, which ties the knot between the
+ *  user-facing type parameter and the type member which couples key and value.
+ */
+final case class AttrOf[A](key: AKey[A], value: A) extends Attr {
+  type Type = A
+  override def toString = s"$key: $value"
+}
+
+object Attr {
+  def apply[A](value: A)(implicit key: AKey[A]): AttrOf[A] = new AttrOf[A](key, value)
+  def unapply(x: Attr): Some[(AKey[x.Type], x.Type)]       = Some(x.pair)
+}
+
+/** A Metadata map, holding any number of well typed Attrs.
+ *  A typed value can be obtained for any key.
+ */
+final class Metadata(attrs: Vector[Attr]) {
+  private[this] val untypedMap                   = attrs.foldLeft(Map[AKey[_], Any]())(_ + _.pair)
+  private def doApply[A]()(implicit z: AKey[A]): A = untypedMap(z).asInstanceOf[A]
+
+  def isEmpty                                = untypedMap.isEmpty
+  def apply[A]()(implicit z: AKey[A]): A     = if (has[A]) doApply[A]() else z.emptyValue
```

Comment:
> I don't think Size(0) is a good way of saying: we don't know because now we can not differentiate between 'we don't know' and 'it's 0'.

You certainly don't use 0 for this because 0 is a legal value which defeats the whole point. It would be like using `Some(0)` to mean `None`. You use the designated empty value. In the world of nonnegative Ints, it's usually -1. If you bring some principle to the process, maybe it's NoSize.